clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -7161,6 +7169,90 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
       FFPrinter.printFreeFunctionDeclaration(K.SyclKernel, ParmListWithNames);
     }
 
+    // Now we handle all structs that contain special types
+    // inside. Their information is contained in StructsWithSpecialTypes of
+    // SemaSYCL.
+    for (ParmVarDecl *Param : K.SyclKernel->parameters()) {
+      if (!Param->getType()->isStructureType())
+        continue;
+      const RecordDecl *Struct =
+          Param->getType()->getAsStructureType()->getDecl();
+      QualType type = Param->getType();
+      if (!S.getStructsWithSpecialType().count(Struct) ||
+          visitedStructWithSpecialType.count(Struct))
+        continue;
+
+      FwdDeclEmitter.Visit(type.getDesugaredType(S.getASTContext()));
+
+      //  this is a struct that contains a special type so its neither a
+      //  special type nor a trivially copyable type. We therefore need to
+      //  explicitly communicate to the runtime that this argument should be
+      //  allowed as a free function kernel argument. We do this by defining
+      //   is_struct_with_special_type to be true.
+      O << "template <>\n";
+      O << "struct "
+           "sycl::ext::oneapi::experimental::detail::is_struct_with_special_"
+           "type<";
+      Policy.SuppressTagKeyword = true;
+      type.print(O, Policy);
+      O << "> {\n";
+      O << " inline static constexpr bool value = true;\n};\n\n";
+      // Now we define the set_arg function for this struct that contains
+      // special types. It takes the handler as an argument so that we can
+      // ultimately call set_arg on the handler for each member
+      // contained in the struct. First, emit needed forward declarations for
+      // all types contained in the struct by doing a depth first search
+      // exploration of the struct. Also collect the offsets of all the
+      // members inside the struct while we're at it so that we can call
+      // set_arg for each one of those.
+      llvm::SmallVector<std::pair<const FieldDecl *, uint64_t>, 8> offsets;
+      llvm::SmallVector<const RecordDecl *, 8> dfs;
+      dfs.emplace_back(Struct);
+      while (!dfs.empty()) {
+        const auto top = dfs.pop_back_val();
+        for (const auto member : top->fields()) {
+          if (isSyclSpecialType(member->getType(), S) ||
+              !member->getType()->isStructureType()) {
+            FwdDeclEmitter.Visit(
+                member->getType().getDesugaredType(S.getASTContext()));
+            offsets.emplace_back(std::make_pair(
+                member, S.getASTContext().getFieldOffset(member) / 8));
```

Comment:
For SYCL 2020 kernels we calculate offsets in `SyclKernelIntHeaderCreator`. Why can't we do that for free function kernels.